src/databricks/labs/remorph/mixins/root_table.py

```diff
@@ -0,0 +1,143 @@
+        if node_name not in self.nodes:
+            self.nodes[node_name] = Node(node_name)
+
+    def add_edge(self, parent_name, child_name):
+        parent_name = parent_name.upper()  # convert to upper case
+        child_name = child_name.upper()  # convert to upper case
+        if parent_name not in self.nodes:
+            self.add_node(parent_name)
+        if child_name not in self.nodes:
+            self.add_node(child_name)
+        self.nodes[parent_name].add_child(self.nodes[child_name])
+
+    def __repr__(self):
+        return str(self.nodes.values())
+
+
+class RootTableIdentifier:
+    def __init__(self, source, input_path):
+        self.source = source
+        self.input_path = input_path
+        self.root_tables = defaultdict(int)
+        self.dag = DAG()
+
+    def generate_lineage(self):
+        # when input is sql file then parse the file
+        if is_sql_file(self.input_path):
+            filename = self.input_path
+            sql_content = read_file(filename)
+            self._parse_sql_content(sql_content, filename)
+            return  # return after processing the file
+
+        # when the input is a directory
+        for filename in get_sql_file(self.input_path):
+            print("***************")
+            print(filename)
+            logger.debug(f"Processing file: {filename}")
+            sql_content = read_file(filename)
+            self._parse_sql_content(sql_content, filename)
+
+    def identify_root_tables(self, level):
+        all_nodes = set(self.dag.nodes.values())
+        child_nodes = {node for n in self.dag.nodes.values() for node in n.children}
+        root_nodes = all_nodes - child_nodes
+
+        # Now we have all root nodes. We need to find the ones at the specified level. We can do this by performing a
+        # breadth-first search (BFS) from each root node and stopping at the specified level.
+
+        root_tables_at_level = set()
+
+        for root_node in root_nodes:
+            queue = [(root_node, 0)]  # The queue for the BFS. Each element is a tuple (node, level).
+            while queue:
+                node, node_level = queue.pop(0)
+
+                if node_level == level:
+                    if not any(parent for parent in node.parents if parent.get_degree() == node_level):
+                        root_tables_at_level.add(node.name)
+                elif node_level < level:
+                    for child in node.children:
+                        queue.append((child, node_level + 1))
+
+        # The list root_tables_at_level now contains all root tables at the specified level.
+        # We can print them or return them as needed.
+
+        return root_tables_at_level
+
+    @staticmethod
+    def _find_root_tables(expression) -> str:
+        for table in expression.find_all(exp.Table, bfs=False):
+            return str(table.name).upper()
+
+    def _parse_sql_content(self, sql_content, file_name):
```

Comment:
```suggestion
    def _parse_sql_content(self, sql_content, file_name, dag: DAG):
```
this way we're avoiding shared mutable state and prevent future multi-threading bugs

Reply:
Done